React/Base/RCTComponentEvent.h

```diff
@@ -0,0 +1,19 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <React/RCTEventDispatcher.h>
+
+/**
+ * Event that is dispatched to React components.
+ */
+@interface RCTComponentEvent : NSObject<RCTEvent>
+
+- (instancetype)initWithName:(NSString *)name body:(NSDictionary *)body;
```

Comment:
Hmm, why we initialize it with untyped NSDictionary? I thought whole point of having RCTEvent objects is typesafty and explicitly.

How do you see the perpose of this class? Something simple which should replace all current untyped events? Should we have it?

Reply:
I think the main use of RCTEvent is to support event coalescing. RCTComponentEvent is a generic event that is sent to a component so it could be something like a touch or scroll event and those can all have different bodies. It is also used by event blocks created by view managers (RCTDirectEventBlock, RCTBubblingEventBlock). If someone want to make a type safe event I think the best way is to create a new class that implements RCTEvent or subclass RCTComponentEvent. An example of this is https://github.com/facebook/react-native/blob/master/React/Views/RCTScrollView.m#L26

Reply:
Yeah!
(How can it be used by `RCT*EventBlock` if you just added this?)
So, in my mind we have to advice to product developers use `RCT*EventBlock` or Custom implementation of `RCTEvent`. We have to discourage them to use this untyped and generic RCTComponentEvent. But, it is okay to use it as a temporary solution for migration from `sendEventWithName`.

* `RCT*EventBlock` Is easy to use.
* Custom RCTEvent implementation is powerful and semantically explicit.

This should be the most important point of this change. Seems that Nick Lockwood's original idea when he started this migration awhile ago.

Do you agree with this vision? Is this reasonable?

Reply:
(Oh, I understood where we use it as a bearer for `RCTEvent*Block`, that's great! Fantastic!)

Reply:
Yea this class is used internally and could also be used as an easy migration from sendEventWithName. Other events should just implement RCTEvent